src/util/locktime.rs

```diff
@@ -0,0 +1,516 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Provides type `LockTime` that implements the logic around nLockTime/OP_CHECKLOCKTIMEVERIFY.
+//!
+
+use core::{mem, fmt};
+use core::cmp::{PartialOrd, Ordering};
+use core::convert::TryFrom;
+
+use crate::consensus::encode::{self, Decodable, Encodable};
+use crate::io;
+
+/// The Threshold for deciding whether a lock time value is interpreted as height or time (see [Bitcoin Core]).
+///
+/// [Bitcoin Core]: https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+pub const LOCKTIME_THRESHOLD: u32 = 500_000_000;
+
+/// A transaction with nLocktime==0 is able to be included immediately in any block.
+pub const ZERO: LockTime = LockTime::Blocks(Height(0));
```

Comment:
I believe it's idiomatic to have this as associated constant of `LockTime`.

Reply:
I'm not sure what to do about this. AFAICT associated consts are not available for enums (unless I'm missing something).

Slightly off topic but in relation to the identifier 'zero', as discussed above, it requires domain knowledge to understand. I was uneasy about adding `ZERO`. We could possibly use `LockTime::INCLUDE_IMMEDIATELY`?

Reply:
I think zero is much clearer. `INCLUDE_IMMEDIATELY` sounds like a magic constant, but then it isn't referenced anywhere in any BIP, nor is even the fact that "zero" means "include immediately" since this fact is well-known.

Reply:
> I'm not sure what to do about this. AFAICT associated consts are not available for enums (unless I'm missing something).

[They are](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1f0d2a52a5a2fd421c56ba85bde02d88)

> Slightly off topic but in relation to the identifier 'zero', as discussed above, it requires domain knowledge to understand. I was uneasy about adding `ZERO`. We could possibly use `LockTime::INCLUDE_IMMEDIATELY`?

More like `LockTime::ALWAYS_SATISFIED`. But yes, not being `ZERO` may confuse experienced devs while being `ZERO` will be harder for inexperienced devs.

Reply:
I really don't believe anybody's life will be made easier by us creating our own new name for zero.

It would also defeat the purpose of having this constant if it were longer to type than `LockTime::from(0)`.